targets/TARGET_STM/qspi_api.c

```diff
@@ -0,0 +1,260 @@
+        case QSPI_CFG_BUS_SINGLE:
+            st_command->AddressMode = QSPI_ADDRESS_1_LINE;
+            break;
+        case QSPI_CFG_BUS_DUAL:
+            st_command->AddressMode = QSPI_ADDRESS_2_LINES;
+            break;
+        case QSPI_CFG_BUS_QUAD:
+            st_command->AddressMode = QSPI_ADDRESS_4_LINES;
+            break;
+        default:
+            st_command->AddressMode = QSPI_ADDRESS_NONE;
+            break;
+    }
+
+    if (command->address.disabled == true) {
+        st_command->AddressMode = QSPI_ADDRESS_NONE;
+    }
+
+    switch (command->alt.bus_width) {
+        case QSPI_CFG_BUS_SINGLE:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_1_LINE;
+            break;
+        case QSPI_CFG_BUS_DUAL:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_2_LINES;
+            break;
+        case QSPI_CFG_BUS_QUAD:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_4_LINES;
+            break;
+        default:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_NONE;
+            break;
+    }
+
+    if (command->alt.disabled == true) {
+        st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_NONE;
+    }
+
+    switch (command->data.bus_width) {
+        case QSPI_CFG_BUS_SINGLE:
+            st_command->DataMode = QSPI_DATA_1_LINE;
+            break;
+        case QSPI_CFG_BUS_DUAL:
+            st_command->DataMode = QSPI_DATA_2_LINES;
+            break;
+        case QSPI_CFG_BUS_QUAD:
+            st_command->DataMode = QSPI_DATA_4_LINES;
+            break;
+        default:
+            st_command->DataMode = QSPI_DATA_NONE;
+            break;
+    }
+}
+
+
+qspi_status_t qspi_init(qspi_t *obj, PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName ssel, uint32_t hz, uint8_t mode)
+{
+    // Enable interface clock for QSPI
+     __HAL_RCC_QSPI_CLK_ENABLE();
+
+    // Reset QSPI
+    __HAL_RCC_QSPI_FORCE_RESET();
+    __HAL_RCC_QSPI_RELEASE_RESET();
+
+    // Set default QSPI handle values
+    obj->handle.Init.ClockPrescaler = 1;
+    obj->handle.Init.FifoThreshold = 4;
+    obj->handle.Init.SampleShifting = QSPI_SAMPLE_SHIFTING_HALFCYCLE;
+    obj->handle.Init.FlashSize = POSITION_VAL(QSPI_FLASH_SIZE_DEFAULT) - 1;
+    obj->handle.Init.ChipSelectHighTime = QSPI_CS_HIGH_TIME_2_CYCLE;
+    obj->handle.Init.ClockMode = QSPI_CLOCK_MODE_0;
+    obj->handle.Init.FlashID = QSPI_FLASH_ID_1;
+    obj->handle.Init.DualFlash = QSPI_DUALFLASH_DISABLE;
+
+    obj->handle.Init.ClockMode = mode == 0 ? QSPI_CLOCK_MODE_0 : QSPI_CLOCK_MODE_3;
+
+    QSPIName qspi_data_first = (SPIName)pinmap_merge(io0, io1);
+    QSPIName qspi_data_second = (SPIName)pinmap_merge(io1, io2);
+    QSPIName qspi_data_third = (SPIName)pinmap_merge(io2, io3);
+
+    if (qspi_data_first != qspi_data_second || qspi_data_second != qspi_data_third ||
+        qspi_data_first != qspi_data_third) {
+        return QSPI_STATUS_INVALID_PARAMETER;
+    }
```

Comment:
Hi @0xc0170 
pinmap_merge is supposed to check that we are using pins from the same HW module. It takes the HW name as input parameters and not directly pin names.
The correct implementation would look like
```
    QSPIName qspiio0name = (QSPIName)pinmap_peripheral(io0, PinMap_QSPI_DATA);
    QSPIName qspiio1name = (QSPIName)pinmap_peripheral(io1, PinMap_QSPI_DATA);
    QSPIName qspiio2name = (QSPIName)pinmap_peripheral(io2, PinMap_QSPI_DATA);
    QSPIName qspiio3name = (QSPIName)pinmap_peripheral(io3, PinMap_QSPI_DATA);
    QSPIName qspiclkname = (QSPIName)pinmap_peripheral(sclk, PinMap_QSPI_SCLK);
    QSPIName qspisselname = (QSPIName)pinmap_peripheral(ssel, PinMap_QSPI_SSEL);

    QSPIName qspi_data_first = (QSPIName)pinmap_merge(qspiio0name, qspiio1name);
    QSPIName qspi_data_second = (QSPIName)pinmap_merge(qspiio2name, qspiio3name);
    QSPIName qspi_data_third = (QSPIName)pinmap_merge(qspiclkname, qspisselname);

    if (qspi_data_first != qspi_data_second || qspi_data_second != qspi_data_third ||
        qspi_data_first != qspi_data_third) {
        return QSPI_STATUS_INVALID_PARAMETER;
    }

```